### ISSUE TITLE
Warmup question

### ISSUE THREAD

**Issue by dariogonle** [ultralytics/yolov5#3156]
## ❔Question

Hi,

 I'm playing a little bit with the hyperparameters and now I'm focusing in warmup. I modify train.py to log the LR used in each epoch and I understand how LR changes among them, but if I change the warmup_epochs hyperparameter no effect is shown. I think that warmup is used to train with a high LR (close to 0) for the firsts epochs and then use lr0. That would mean that the LR evolution should follow this curve.

![image](https://user-images.githubusercontent.com/80393891/118150521-f0f60c80-b412-11eb-8a77-41c7f38a01ee.png)

But when I log the LR in each epoch, I start viewing lr0 and the it goes smaller, but if I set up  warmup_epochs = 3 and warmup_bias_lr = 0.1, I think it should start with a LR of 0.1 at epoch 0, change to 0.01 (lr0) at epoch 3 and then start to decrease untill 0.002 (lrf)

Am I right?

I would also like to know if it is possible to establish a LR like the blue curve shown in the image. I mean if it is already coded, I guess I could do it by myself.

Thank you.

**Comment by glenn-jocher** [comment]
@dariogonle the automatic LR schedulers built in to YOLOv5 are one cycle LR (default) and linear (with `--linear-lr` flag), both of which first obey the warmup hyperparameters, though you can replace these with any custom scheduler as well by modifying the train.py code. The warmup slowly updates the LR from the warmup LR0 to the training LR defined by the LR scheduler. They meet at the end of epoch 2 under the default hyps. You can use plot_lr() to preview what your LR will look like (minus warmup):
https://github.com/ultralytics/yolov5/blob/06372b1465f5a58463bf8c32bdf65fc679c17ebf/train.py#L134-L142

You can see the 3 parameter group LRs using TensorBoard or W&B logging:
<img width="978" alt="Screen Shot 2021-05-13 at 6 10 16 PM" src="https://user-images.githubusercontent.com/26833433/118154068-e9d0fd80-b416-11eb-8bd4-fa24d682fcf7.png">

**Comment by dariogonle** [comment]
Thank you @glenn-jocher. I know this may not have to deal with YOLOv5 but why are 3 learning rates used? What's the meaning of lr0, lr1 and lr2?

**Comment by glenn-jocher** [comment]
@dariogonle see train.py:
https://github.com/ultralytics/yolov5/blob/06372b1465f5a58463bf8c32bdf65fc679c17ebf/train.py#L109-L123

**Comment by dariogonle** [comment]
Thank you again @glenn-jocher. I've been reading a little bit about all this things and I have a question. In some papers I read that a low warmup_lr is used (e.g. 0.001), then a higher initial lr (e.g. 0.01) and then it is decreased again (e.g. 0.002 at the last epoch). The following picture shows this behabiour:
![image](https://user-images.githubusercontent.com/80393891/118243866-37914880-b49f-11eb-9dbc-408c7c1a610f.png)


But if I take a look at hyp.yaml, I see that the warmup_bias_lr is set to 0.1, much bigger than lr0, what means that the behaviour is the opposite to what I explain above.
https://github.com/ultralytics/yolov5/blob/06372b1465f5a58463bf8c32bdf65fc679c17ebf/data/hyp.scratch.yaml#L6-L12

The behaviour is explained in this picture.
![image](https://user-images.githubusercontent.com/80393891/118244138-7e7f3e00-b49f-11eb-967c-7ddc34ef9dac.png)


From your experience what's the best method?

**Comment by glenn-jocher** [comment]
@dariogonle different strategies are applicable to different parameter groups, hence the current implementation.
